lib/SIL/Utils/DynamicCasts.cpp

```diff
@@ -247,26 +324,12 @@ bool SILDynamicCastInst::isRCIdentityPreserving() const {
   // would get confused and might eliminate a retain of such an object
   // completely.
   SILFunction &f = *getFunction();
-  if (getSourceLoweredType().isTrivial(f) != getTargetLoweredType().isTrivial(f))
-    return false;
-
-  CanType source = getSourceFormalType();
-  CanType target = getTargetFormalType();
-
-  // An existential may be holding a reference to a bridgeable struct.
-  // In this case, ARC on the existential affects the refcount of the container
-  // holding the struct, not the class to which the struct is bridged.
-  // Therefore, don't assume RC identity when casting between existentials and
-  // classes (and also between two existentials).
-  if (canBeExistential(source) &&
-      (canBeClass(target) || canBeExistential(target)))
-    return false;
-
-  // And vice versa.
-  if (canBeClass(source) && canBeExistential(target))
-    return false;
-
-  return true;
+  if (getSourceLoweredType().isTrivial(f)
+      && getTargetLoweredType().isTrivial(f)) {
+    return true;
+  }
+  return doesCastPreserveOwnershipForType(getSourceFormalType())
+    && doesCastPreserveOwnershipForType(getTargetFormalType());
```

Comment:
If I understand this code correctly, it's more conservative than what you described in the comment with (A), (B1) and (B2). Maybe you add a short comment for these conditions to relate them to A, B1, B2.

Specifically, does this include casting a class existential to a class which conforms to that protocol?
I'm asking because this is an important use case in libswift for fast casting (https://github.com/apple/swift/pull/41784).

Reply:
From the comments:
``` 
// A type's ownership is preserved by a dynamic cast if it is neither
//   (A) a potentially bridged value
//       (isPotentiallyBridgedValueOrExistentialType())
// nor
//   (B) potentially wrapped in a transparent type
//       (Type::isPotentiallyAnyObject())
```

A non-AnyObject class existential meets both requirements, as does a nominal class. I'll make sure there's some kind of test for that case.

Reply:
@eeckstein I've lost track of the number of rewrites, but with the last attempt I think it's reaching the point where anyone can understand it. Still working on tests for all the interesting cases